### PR TITLE
Add create record and extract type code actions

### LINKED ISSUE #32093: New code action to extract anonymous records into record types
**Description:**
Consider the following scenario:

```ballerina
service / on http:Listener(9090) {
   resource function get greeting() returns record {*http:Ok; json body} {
       return {};
   }
}
```
With the new codeaction, we can extract out the `record {*http:Ok; json body}` anonymous record to
```
type GreetingResponse record {*http:Ok; json body};
```
and use it in the return type descriptor. Similarly there can be many use cases for this.

**Describe your problem(s)**
See description

**Describe your solution(s)**
See description

### LINKED ISSUE #35333: Code action to generate record types
Usually when working with services we start with defining the service interface. Consider the below program.

Screen Shot 2022-03-04 at 4 17 46 PM

Here once I type `Payment`, I believe there should be a code action to generate the following.

Screen Shot 2022-03-04 at 4 23 51 PM

Anyway, I think this is a generic feature not specific to services.

### PR DESCRIPTION
**Pull request by IMS94**
## Purpose
$subject

Fixes #35333
Fixes #32093

## Samples
extract_type_codeaction

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractTypeCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                context.fileUri(), CodeActionKind.RefactorExtract);
+        return List.of(codeAction);
+    }
+
+    @Override
+    public String getName() {
+        return "ExtractType";
+    }
+
+    private static String generateTypeName(String prefix, Set<String> visibleSymbolNames) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractTypeCodeAction.java:108]: Shall we add this method  or reuse one of the methods in CommonUtil to generate the type name
https://github.com/ballerina-platform/ballerina-lang/blob/02411da0cf01ac4f76c768ed911728b0e0ad115b/language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java#L718

**Review by malinthar — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        Node tlNode = node.get();
+        boolean isReturnType = false;
+        while (tlNode.parent().kind() != SyntaxKind.MODULE_PART) {
+            if (tlNode.kind() == SyntaxKind.RETURN_TYPE_DESCRIPTOR) {
+                isReturnType = true;
+            }
+            tlNode = tlNode.parent();
+        }
+        
+        Range range = new Range(CommonUtil.toPosition(tlNode.lineRange().startLine()),
```
[the remark this answers is not in the document]
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java:72]: Agreed. +1. Will come up with a way.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import org.wso2.ballerinalang.compiler.util.Name;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Code action to create a record for an unknown type.
+ *
+ * @since 2201.0.3
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java:42]: Version should be changed to 2201.1.1 ?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractTypeCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The code action to extract anonymous records into type defs.
+ *
+ * @since 2201.0.3
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractTypeCodeAction.java:51]: Version

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateTypeCodeActionTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.ballerinalang.langserver.commons.workspace.WorkspaceDocumentException;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test class to test the functionality of the create type code action.
+ *
+ * @since 2201.0.x
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CreateTypeCodeActionTest.java:27]: version

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ExtractTypeCodeActionTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import org.ballerinalang.langserver.commons.workspace.WorkspaceDocumentException;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test class to test the functionality of the extract type code action.
+ *
+ * @since 2201.0.x
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/ExtractTypeCodeActionTest.java:27]: Version

**Review by malinthar — COMMENTED** [review]
LGTM! Some minor version changes

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java (hunk 3 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateTypeCodeAction.java]: Yes. Done!

**Review by malinthar — APPROVED** [review]
LGTM!

**Pull request merged**